In [1]:
import pandas as pd 
import numpy as nd

In [4]:
df=pd.read_csv('telecom_customer_churn.csv')

In [35]:
rowColumn = df.shape
print(f"Total number of rows and coloum: {rowColumn}")

Total number of rows and coloum: (6923, 38)


In [26]:
df

Customer ID  Gender  Age Married  Number of Dependents          City  \
0     0002-ORFBO  Female   37     Yes                     0  Frazier Park   
1     0003-MKNFE    Male   46      No                     0      Glendale   
2     0004-TLHLJ    Male   50      No                     0    Costa Mesa   
3     0011-IGKFF    Male   78     Yes                     0      Martinez   
4     0013-EXCHZ  Female   75     Yes                     0     Camarillo   
...          ...     ...  ...     ...                   ...           ...   
7038  9987-LUTYD  Female   20      No                     0       La Mesa   
7039  9992-RRAMN    Male   40     Yes                     0     Riverbank   
7040  9992-UJOEL    Male   22      No                     0           Elk   
7041  9993-LHIEB    Male   21     Yes                     0  Solana Beach   
7042  9995-HOTOH    Male   36     Yes                     0   Sierra City   

      Zip Code   Latitude   Longitude  Number of Referrals  ...  \
0        93225  34.827662 -118.999073                    2  ...   
1        91206  34.162515 -118.203869                    0  ...   
2        92627  33.645672 -117.922613                    0  ...   
3        94553  38.014457 -122.115432                    1  ...   
4        93010  34.227846 -119.079903                    3  ...   
...        ...        ...         ...                  ...  ...   
7038     91941  32.759327 -116.997260                    0  ...   
7039     95367  37.734971 -120.954271                    1  ...   
7040     95432  39.108252 -123.645121                    0  ...   
7041     92075  33.001813 -117.263628                    5  ...   
7042     96125  39.600599 -120.636358                    1  ...   

       Payment Method Monthly Charge Total Charges  Total Refunds  \
0         Credit Card          65.60        593.30           0.00   
1         Credit Card          -4.00        542.40          38.33   
2     Bank Withdrawal          73.90        280.85           0.00   
3     Bank Withdrawal          98.00       1237.85           0.00   
4         Credit Card          83.90        267.40           0.00   
...               ...            ...           ...            ...   
7038      Credit Card          55.15        742.90           0.00   
7039  Bank Withdrawal          85.10       1873.70           0.00   
7040      Credit Card          50.30         92.75           0.00   
7041      Credit Card          67.85       4627.65           0.00   
7042  Bank Withdrawal          59.00       3707.60           0.00   

     Total Extra Data Charges Total Long Distance Charges Total Revenue  \
0                           0                      381.51        974.81   
1                          10                       96.21        610.28   
2                           0                      134.60        415.45   
3                           0                      361.66       1599.51   
4                           0                       22.14        289.54   
...                       ...                         ...           ...   
7038                        0                      606.84       1349.74   
7039                        0                      356.40       2230.10   
7040                        0                       37.24        129.99   
7041                        0                      142.04       4769.69   
7042                        0                        0.00       3707.60   

      Customer Status   Churn Category                   Churn Reason  
0              Stayed              NaN                            NaN  
1              Stayed              NaN                            NaN  
2             Churned       Competitor  Competitor had better devices  
3             Churned  Dissatisfaction        Product dissatisfaction  
4             Churned  Dissatisfaction            Network reliability  
...               ...              ...                            ...  
7038           Stayed              NaN   

In [8]:
Rduplicate = df.duplicated().sum()
print(f"Number of duplicate rows: {Rduplicate}")

Number of duplicate rows: 0


In [13]:
missingVal = df.isnull().sum().sum()
print(f"Missing value in DATA SET:,{missingVal}")

Missing value in DATA SET:,30849


In [21]:
print(df.isnull().sum())

Customer ID                             0
Gender                                  0
Age                                     0
Married                                 0
Number of Dependents                    0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Number of Referrals                     0
Tenure in Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges     682
Multiple Lines                        682
Internet Service                        0
Internet Type                        1526
Avg Monthly GB Download              1526
Online Security                      1526
Online Backup                        1526
Device Protection Plan               1526
Premium Tech Support                 1526
Streaming TV                         1526
Streaming Movies                  

In [14]:
df_cleaned = df.dropna()

In [25]:
print(df_cleaned)

     Customer ID  Gender  Age Married  Number of Dependents             City  \
2     0004-TLHLJ    Male   50      No                     0       Costa Mesa   
3     0011-IGKFF    Male   78     Yes                     0         Martinez   
54    0094-OIFMO  Female   78      No                     1  North Hollywood   
67    0112-QWPNC    Male   61     Yes                     0         Valyermo   
77    0125-LZQXK    Male   42      No                     0     Big Oak Flat   
...          ...     ...  ...     ...                   ...              ...   
7007  9940-RHLFB  Female   53      No                     0     Sunset Beach   
7010  9944-HKVVB  Female   55      No                     0        King City   
7012  9947-OTFQU    Male   65      No                     0           Covina   
7036  9985-MWVIX  Female   53      No                     0             Hume   
7039  9992-RRAMN    Male   40     Yes                     0        Riverbank   

      Zip Code   Latitude   Longitude  

In [27]:
df_cleaned_columns = df.dropna(axis=1)

In [28]:
print(df_cleaned)

     Customer ID  Gender  Age Married  Number of Dependents             City  \
2     0004-TLHLJ    Male   50      No                     0       Costa Mesa   
3     0011-IGKFF    Male   78     Yes                     0         Martinez   
54    0094-OIFMO  Female   78      No                     1  North Hollywood   
67    0112-QWPNC    Male   61     Yes                     0         Valyermo   
77    0125-LZQXK    Male   42      No                     0     Big Oak Flat   
...          ...     ...  ...     ...                   ...              ...   
7007  9940-RHLFB  Female   53      No                     0     Sunset Beach   
7010  9944-HKVVB  Female   55      No                     0        King City   
7012  9947-OTFQU    Male   65      No                     0           Covina   
7036  9985-MWVIX  Female   53      No                     0             Hume   
7039  9992-RRAMN    Male   40     Yes                     0        Riverbank   

      Zip Code   Latitude   Longitude  

In [30]:
df.describe()

Age  Number of Dependents      Zip Code     Latitude  \
count  7043.000000           7043.000000   7043.000000  7043.000000   
mean     46.509726              0.468692  93486.070567    36.197455   
std      16.750352              0.962802   1856.767505     2.468929   
min      19.000000              0.000000  90001.000000    32.555828   
25%      32.000000              0.000000  92101.000000    33.990646   
50%      46.000000              0.000000  93518.000000    36.205465   
75%      60.000000              0.000000  95329.000000    38.161321   
max      80.000000              9.000000  96150.000000    41.962127   

         Longitude  Number of Referrals  Tenure in Months  \
count  7043.000000          7043.000000       7043.000000   
mean   -119.756684             1.951867         32.386767   
std       2.154425             3.001199         24.542061   
min    -124.301372             0.000000          1.000000   
25%    -121.788090             0.000000          9.000000   
50%    -119.595293             0.000000         29.000000   
75%    -117.969795             3.000000         55.000000   
max    -114.192901            11.000000         72.000000   

       Avg Monthly Long Distance Charges  Avg Monthly GB Download  \
count                        6361.000000              5517.000000   
mean                           25.420517                26.189958   
std                            14.200374                19.586585   
min                             1.010000                 2.000000   
25%                            13.050000                13.000000   
50%                            25.690000                21.000000   
75%                            37.680000                30.000000   
max                            49.990000                85.000000   

       Monthly Charge  Total Charges  Total Refunds  Total Extra Data Charges  \
count     7043.000000    7043.000000    7043.000000               7043.000000   
mean        63.596131    2280.381264       1.962182                  6.860713   
std         31.204743    2266.220462       7.902614                 25.104978   
min        -10.000000      18.800000       0.000000                  0.000000   
25%         30.400000     400.150000       0.000000                  0.000000   
50%         70.050000    1394.550000       0.000000                  0.000000   
75%         89.750000    3786.600000       0.000000                  0.000000   
max        118.750000    8684.800000      49.790000                150.000000   

       Total Long Distance Charges  Total Revenue  
count                  7043.000000    7043.000000  
mean                    749.099262    3034.379056  
std                     846.660055    2865.204542  
min                       0.000000      21.360000  
25%                      70.545000     605.610000  
50%                     401.440000    2108.640000  
75%                    1191.100000    4801.145000  
max                    3564.720000   11979.340000

In [31]:
df = df[df['Monthly Charge'] >= 0]

In [32]:
df

Customer ID  Gender  Age Married  Number of Dependents          City  \
0     0002-ORFBO  Female   37     Yes                     0  Frazier Park   
2     0004-TLHLJ    Male   50      No                     0    Costa Mesa   
3     0011-IGKFF    Male   78     Yes                     0      Martinez   
4     0013-EXCHZ  Female   75     Yes                     0     Camarillo   
5     0013-MHZWF  Female   23      No                     3      Midpines   
...          ...     ...  ...     ...                   ...           ...   
7038  9987-LUTYD  Female   20      No                     0       La Mesa   
7039  9992-RRAMN    Male   40     Yes                     0     Riverbank   
7040  9992-UJOEL    Male   22      No                     0           Elk   
7041  9993-LHIEB    Male   21     Yes                     0  Solana Beach   
7042  9995-HOTOH    Male   36     Yes                     0   Sierra City   

      Zip Code   Latitude   Longitude  Number of Referrals  ...  \
0        93225  34.827662 -118.999073                    2  ...   
2        92627  33.645672 -117.922613                    0  ...   
3        94553  38.014457 -122.115432                    1  ...   
4        93010  34.227846 -119.079903                    3  ...   
5        95345  37.581496 -119.972762                    0  ...   
...        ...        ...         ...                  ...  ...   
7038     91941  32.759327 -116.997260                    0  ...   
7039     95367  37.734971 -120.954271                    1  ...   
7040     95432  39.108252 -123.645121                    0  ...   
7041     92075  33.001813 -117.263628                    5  ...   
7042     96125  39.600599 -120.636358                    1  ...   

       Payment Method Monthly Charge Total Charges  Total Refunds  \
0         Credit Card          65.60        593.30            0.0   
2     Bank Withdrawal          73.90        280.85            0.0   
3     Bank Withdrawal          98.00       1237.85            0.0   
4         Credit Card          83.90        267.40            0.0   
5         Credit Card          69.40        571.45            0.0   
...               ...            ...           ...            ...   
7038      Credit Card          55.15        742.90            0.0   
7039  Bank Withdrawal          85.10       1873.70            0.0   
7040      Credit Card          50.30         92.75            0.0   
7041      Credit Card          67.85       4627.65            0.0   
7042  Bank Withdrawal          59.00       3707.60            0.0   

     Total Extra Data Charges Total Long Distance Charges Total Revenue  \
0                           0                      381.51        974.81   
2                           0                      134.60        415.45   
3                           0                      361.66       1599.51   
4                           0                       22.14        289.54   
5                           0                      150.93        722.38   
...                       ...                         ...           ...   
7038                        0                      606.84       1349.74   
7039                        0                      356.40       2230.10   
7040                        0                       37.24        129.99   
7041                        0                      142.04       4769.69   
7042                        0                        0.00       3707.60   

      Customer Status   Churn Category                   Churn Reason  
0              Stayed              NaN                            NaN  
2             Churned       Competitor  Competitor had better devices  
3             Churned  Dissatisfaction        Product dissatisfaction  
4             Churned  Dissatisfaction            Network reliability  
5              Stayed              NaN                            NaN  
...               ...              ...                            ...  
7038           Stayed              NaN   

In [34]:
rowColumn = df.shape
print(f"Total number of rows and coloum: {rowColumn}")

Ttal number of rows and coloum: (6923, 38)


Before vs After Report

| Feature        | Before Cleaning          | After Cleaning           |
|----------------|--------------------------|--------------------------|
| Shape          | (7043, 38)               | (6923, 38)               |
| Missing Values | 3 columns had NaNs       | 0 columns have NaNs      |
| Duplicates     | 0 duplicates             | 0 duplicates             |
| Outliers       | Monthly Charge = -10     | Removed                  |
